### PR TITLE
Add PropValuDatatype. 

### PR DESCRIPTION
**Pull request by vEpiphyte**
Update XREFs to be better Synapse citizens. No more left-hand data!

Add refs() support for propvalu, supports using xrefs data types right away!

Add generic xref form migration function.  This will migrate all loaded xref form types when the syn migration is run.

### CODE REVIEW

File: synapse/lib/types.py [hunk 1]
```diff
@@ -353,14 +353,15 @@ def _norm_str(self, text, oldval=None):
...
-            self._raiseBadValu(text)
+            self._raiseBadValu(text, mesg='No text left after strip().')
 
         if text[0] != '(':
             return self.tlib.getTypeNorm('guid', text)
 
         vals, off = s_syntax.parse_list(text)
         if off != len(text):
-            self._raiseBadValu(text)
+            self._raiseBadValu(text, off=off, vals=vals,
```
Comment (vertexmc) [inline synapse/lib/types.py]: These additional error message strings are really helpful

File: synapse/lib/types.py [hunk 2]
```diff
@@ -630,6 +637,60 @@ def norm(self, valu, oldval=None):
...
+
+    def _norm_str(self, text, oldval=None):
+
+        text = text.strip()
+
+        if not text:
+            self._raiseBadValu(text, mesg='No text left after strip().')
+
+        if '=' not in text:
+            self._raiseBadValu(text, mesg='Provalu is missing a =')
```
Comment (vertexmc) [inline synapse/lib/types.py]: Propvalu?

File: synapse/lib/types.py [hunk 3]
```diff
@@ -630,6 +637,60 @@ def norm(self, valu, oldval=None):
...
+        if len(valu) != 2:
+            self._raiseBadValu(valu=valu, mesg='PropValu requires two values to norm.')
+
+        prop, valu = valu
+
+        try:
+            nvalu, nsubs = self._getPropNorm(prop, valu, oldval=oldval)
+        except s_common.BadTypeValu as e:
+            logger.exception('Failed to norm PropValu.')
+            self._raiseBadValu(valu, mesg='Unable to norm PropValu', prop=prop)
```
Comment (vertexmc) [inline synapse/lib/types.py]: Do we need the logger above, since the exception will log almost the same thing?

File: synapse/tests/test_types.py [hunk 4]
```diff
@@ -452,29 +457,46 @@ def test_type_str_strip(self):
         self.eq(tlib.getTypeCast('str:lwr', ' ASDF  '), 'asdf')
 
     def test_type_xref(self):
```
Comment (vertexmc) [inline synapse/tests/test_types.py]: Can we add another test where the `xref:strval` internally contains an `=`? Is that allowed?

File: synapse/lib/types.py (hunk 3 above)
Comment (vEpiphyte) [reply synapse/lib/types.py]: We don't have a super-good way to chain the exceptions the way I'd like too with the helper and python2 support required.  This at least logs what the original norm failure was before wrapping it separately.

File: synapse/lib/types.py (hunk 1 above)
Comment (vEpiphyte) [reply synapse/lib/types.py]: :smile:

File: synapse/lib/types.py (hunk 2 above)
Comment (vEpiphyte) [reply synapse/lib/types.py]: fixed

File: synapse/tests/test_types.py (hunk 4 above)
Comment (vEpiphyte) [reply synapse/tests/test_types.py]: that's totally allowed so i've added a test to do that!

File: synapse/lib/types.py (hunk 3 above)
Comment (vertexmc) [reply synapse/lib/types.py]: ok that's fine then

File: synapse/tests/test_types.py (hunk 4 above)
Comment (vertexmc) [reply synapse/tests/test_types.py]: thanks!

File: synapse/models/syn.py [hunk 5]
```diff
@@ -140,10 +143,21 @@ def _revModl201709191412(self):
...
                 iden = node[0]
                 srcvtype = node[1].get(xtyp)
-                xrefprop = '{}:xref:prop'.format(ntyp)
+                if srcvtype is None:
+                    # This is expensive node level introspection :(
+                    for prop, valu in s_tufo.props(node).items():
+                        if prop.startswith('xref:'):
+                            form = prop.split('xref:', 1)[1]
+                            if self.core.isTufoForm(form):
+                                srcvtype = form
```
Comment (vertexmc) [inline synapse/models/syn.py]: should we break after finding a `srcvtype`?

Comment (vEpiphyte) [reply synapse/models/syn.py]: Yeah we should

Comment (vEpiphyte) [reply synapse/models/syn.py]: fixed

**Pull request merged**